File: TourismAgency/src/main/java/repository/database/ReservationDbRepository.java
package repository.database;

import domain.Reservation;
import domain.Trip;
import domain.validators.ValidationException;
import domain.validators.Validator;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import repository.RepoException;
import repository.ReservationRepository;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ReservationDbRepository implements ReservationRepository {
    private final JdbcUtils dbUtils;
    private final Validator<Reservation> reservationValidator;

    private static final Logger logger= LogManager.getLogger();

    public ReservationDbRepository(Properties props, Validator<Reservation> reservationValidator){
        this.reservationValidator=reservationValidator;
        logger.info("Initializing ReservationRepository with properties: {} ",props);
        dbUtils=new JdbcUtils(props);
    }
    @Override
    public Reservation findOne(Long aLong) {
        logger.traceEntry("finding reservation with id {} ",aLong);
        Connection con=dbUtils.getConnection();

        try(PreparedStatement preStmt=con.prepareStatement("select a.idr,a.client,a.telefon,a.tickets,b.idT,b.place,b.transport,b.price,b.tickets as tick,b.freetickets,b.datat " +
                "from reservation a inner join trip b " +
                "on a.tripid=b.idT where idr=?")){
            preStmt.setLong(1,aLong);
            try(ResultSet result=preStmt.executeQuery()) {
                if (result.next()) {
                    long id = result.getLong("idr");
                    String client = result.getString("client");
                    String tel = result.getString("telefon");
                    int tickets = result.getInt("tickets");
                    long idT = result.getLong("idt");
                    String place = result.getString("place");
                    String transport = result.getString("transport");
                    LocalDateTime dataOra = result.getTimestamp("datat").toLocalDateTime();
                    LocalDate data = dataOra.toLocalDate();
                    LocalTime hour = dataOra.toLocalTime();
                    double price = result.getDouble("price");
                    int ticketsTrip = result.getInt("tick");
                    int freeTickets = result.getInt("freetickets");
                    Trip trip = new Trip(place,transport,data,hour,price,ticketsTrip,freeTickets);
                    trip.setId(idT);

                    Reservation reservation = new Reservation(client,tel,trip,tickets);
                    reservation.setId(id);
                    logger.traceExit(reservation);
                    return reservation;
                }
            }
        }catch (SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit("No reservation found with id {}", aLong);

        return null;
    }

    @Override
    public Iterable<Reservation> findAll() {
        Connection con=dbUtils.getConnection();
        List<Reservation> reservationList=new ArrayList<>();
        try(PreparedStatement preStmt=con.prepareStatement("select a.idr,a.client,a.telefon,a.tickets,b.idT,b.place,b.transport,b.price,b.tickets as tick,b.freetickets,b.datat " +
                "from reservation a inner join trip b " +
                "on a.tripid=b.idT")) {
            try(ResultSet result=preStmt.executeQuery()) {
                while (result.next()) {
                    long id = result.getLong("idr");
                    String client = result.getString("client");
                    String tel = result.getString("telefon");
                    int tickets = result.getInt("tickets");
                    long idT = result.getLong("idt");
                    String place = result.getString("place");
                    String transport = result.getString("transport");
                    LocalDateTime dataOra = result.getTimestamp("datat").toLocalDateTime();
                    LocalDate data = dataOra.toLocalDate();
                    LocalTime hour = dataOra.toLocalTime();
                    double price = result.getDouble("price");
                    int ticketsTrip = result.getInt("tick");
                    int freeTickets = result.getInt("freetickets");
                    Trip trip = new Trip(place,transport,data,hour,price,ticketsTrip,freeTickets);
                    trip.setId(idT);

                    Reservation reservation = new Reservation(client,tel,trip,tickets);
                    reservation.setId(id);
                    reservationList.add(reservation);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB "+e);
        }
        logger.traceExit(reservationList);
        return reservationList;
    }

    @Override
    public Reservation save(Reservation entity) throws ValidationException, RepoException {

        reservationValidator.validate(entity);

        logger.traceEntry("saving reservation {} ",entity);
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("insert into reservation(client,telefon,tripid,tickets) values (?,?,?,?) returning idr")){
                //preStmt.setDouble(1,entity.getId());
            preStmt.setString(1,entity.getClient());
            preStmt.setString(2,entity.getTelefon());
            preStmt.setLong(3,entity.getTripId().getId());
            preStmt.setInt(4,entity.getNrTick());

            try(ResultSet result=preStmt.executeQuery()) {
                if (result.next()) {
                    long id = result.getLong("idr");
                    entity.setId(id);

                }}
        }catch (SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit();
        return null;
    }

    @Override
    public Reservation delete(Long aLong) throws RepoException {
        if (aLong == null)
            throw new RepoException("Repository exception: id must be not null!\n");

        logger.traceEntry("deleting reservation agent with {}",aLong);
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("delete from reservation where idr=?")){
            preStmt.setLong(1,aLong);
            int result=preStmt.executeUpdate();
        }catch (SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        logger.traceExit();
        return null;
    }

    @Override
    public Reservation update(Reservation entity) throws RepoException, ValidationException {
        if (entity == null)
            throw new RepoException("Repository exception: id must be not null!\n");
        reservationValidator.validate(entity);
        if (findOne(entity.getId()) == null) {
            return entity;
        } else{
            logger.traceEntry("updating reservation {} ",entity);
            Connection con=dbUtils.getConnection();
            try(PreparedStatement preStmt=con.prepareStatement("update reservation set client=?,telefon=?,tripid=?,tickets=? where idr=?")){
                //preStmt.setDouble(1,entity.getId());
                preStmt.setString(1,entity.getClient());
                preStmt.setString(2,entity.getTelefon());
                preStmt.setLong(3,entity.getTripId().getId());
                preStmt.setInt(4,entity.getNrTick());
                preStmt.setLong(5,entity.getId());

                int result=preStmt.executeUpdate();
            }catch (SQLException ex){
                logger.error(ex);
                System.out.println("Error DB "+ex);
            }
            logger.traceExit();
        }
        return null;
    }

    @Override
    public int size() {
        logger.traceEntry();
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("select count(*) as [SIZE] from reservation")) {
            try(ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    logger.traceExit(result.getInt("SIZE"));
                    return result.getInt("SIZE");
                }
            }
        }catch(SQLException ex){
            logger.error(ex);
            System.out.println("Error DB "+ex);
        }
        return 0;
    }

    @Override
    public Iterable<Reservation> findReservationByClient(String client1) {
        Connection con=dbUtils.getConnection();
        List<Reservation> reservationList=new ArrayList<>();
        try(PreparedStatement preStmt=con.prepareStatement("select a.idr,a.client,a.telefon,a.tickets,b.idT,b.place,b.transport,b.price,b.tickets as tick,b.freetickets,b.datat from reservation a inner join trip b on a.tripid=b.idT where client like ?")) {
            preStmt.setString(1,"%"+client1+"%");
            try(ResultSet result=preStmt.executeQuery()) {
                while (result.next()) {
                    long id = result.getLong("idr");
                    String client = result.getString("client");
                    String tel = result.getString("telefon");
                    int tickets = result.getInt("tickets");
                    long idT = result.getLong("idt");
                    String place = result.getString("place");
                    String transport = result.getString("transport");
                    LocalDateTime dataOra = result.getTimestamp("datat").toLocalDateTime();
                    LocalDate data = dataOra.toLocalDate();
                    LocalTime hour = dataOra.toLocalTime();
                    double price = result.getDouble("price");
                    int ticketsTrip = result.getInt("tick");
                    int freeTickets = result.getInt("freetickets");
                    Trip trip = new Trip(place,transport,data,hour,price,ticketsTrip,freeTickets);
                    trip.setId(idT);

                    Reservation reservation = new Reservation(client,tel,trip,tickets);
                    reservation.setId(id);
                    reservationList.add(reservation);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB "+e);
        }
        logger.traceExit(reservationList);
        return reservationList;
    }

    @Override
    public Iterable<Reservation> findReservationByTripID(Long tripID) {
        Connection con=dbUtils.getConnection();
        List<Reservation> reservationList=new ArrayList<>();
        try(PreparedStatement preStmt=con.prepareStatement("select a.idr,a.client,a.telefon,a.tickets,b.idT,b.place,b.transport,b.price,b.tickets as tick,b.freetickets,b.datat from reservation a inner join trip b on a.tripid=b.idT where tripid=?")) {
            preStmt.setLong(1,tripID);
            try(ResultSet result=preStmt.executeQuery()) {
                while (result.next()) {
                    long id = result.getLong("idr");
                    String client = result.getString("client");
                    String tel = result.getString("telefon");
                    int tickets = result.getInt("tickets");
                    long idT = result.getLong("idt");
                    String place = result.getString("place");
                    String transport = result.getString("transport");
                    LocalDateTime dataOra = result.getTimestamp("datat").toLocalDateTime();
                    LocalDate data = dataOra.toLocalDate();
                    LocalTime hour = dataOra.toLocalTime();
                    double price = result.getDouble("price");
                    int ticketsTrip = result.getInt("tick");
                    int freeTickets = result.getInt("freetickets");
                    Trip trip = new Trip(place,transport,data,hour,price,ticketsTrip,freeTickets);
                    trip.setId(idT);

                    Reservation reservation = new Reservation(client,tel,trip,tickets);
                    reservation.setId(id);
                    reservationList.add(reservation);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB "+e);
        }
        logger.traceExit(reservationList);
        return reservationList;
    }
}

File: AgentieTurism/AgentieTurism/repository/ReservationDBRepository.cs
using AgentieTurism.domain;
using System;
using System.Collections.Generic;
using System.Data;
using log4net;

namespace AgentieTurism.repository
{
    class ReservationDBRepository : ReservationRepository
    {
        private static ILog log = LogManager.GetLogger("ReservationDBRepository");

        public ReservationDBRepository()
        {
            log.Info("Creating Reservation DB Repository");
        }

        public void delete(long id)
        {
            log.InfoFormat("Deleting reservation with id {0}", id);
            IDbConnection con = DBUtils.getConnection();

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "delete from reservation where idr=@id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);
                var dataR = comm.ExecuteNonQuery();

                if (dataR == 0)
                    throw new RepositoryException("No reservation deleted!");
                else
                    log.Info("Reservation was deleted succesfully");
            }
        }

        public IEnumerable<Reservation> findAll()
        {
            IDbConnection con = DBUtils.getConnection();
            IList<Reservation> resList = new List<Reservation>();
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select a.idr,a.client,a.telefon,a.tickets,b.idT,b.place,b.transport,b.price,b.tickets,b.freetickets,b.datat from reservation a inner join trip b on a.tripid=b.idT";

                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        long idR = dataR.GetInt64(0);
                        string client = dataR.GetString(1);
                        string telefon = dataR.GetString(2);
                        int tickReserved = dataR.GetInt32(3);

                        long idT = dataR.GetInt64(4);
                        String place = dataR.GetString(5);
                        String transport = dataR.GetString(6);
                        double price = dataR.GetDouble(7);
                        int tickets = dataR.GetInt32(8);
                        int freeTickets = dataR.GetInt32(9);
                        DateTime data = dataR.GetDateTime(10);
                        Trip trip = new Trip(place, transport, data, price, tickets, freeTickets);
                        trip.ID = idT;
                        Reservation reservation = new Reservation(client, telefon, trip, tickReserved);
                        reservation.ID = idR;

                        resList.Add(reservation);
                    }
                }
            }

            return resList;
        }

        public Reservation findOne(long id)
        {
            log.InfoFormat("Entering findOne with value {0}", id);
            IDbConnection con = DBUtils.getConnection();

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select a.idr,a.client,a.telefon,a.tickets,b.idT,b.place,b.transport,b.price,b.tickets,b.freetickets,b.datat" +
                    " from reservation a" +
                    " inner join trip b on a.tripid=b.idT" +
                    " where idr=@id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        long idR = dataR.GetInt64(0);
                        string client = dataR.GetString(1);
                        string telefon = dataR.GetString(2);
                        int tickReserved = dataR.GetInt32(3);

                        long idT = dataR.GetInt64(4);
                        String place = dataR.GetString(5);
                        String transport = dataR.GetString(6);
                        double price = dataR.GetDouble(7);
                        int tickets = dataR.GetInt32(8);
                        int freeTickets = dataR.GetInt32(9);
                        DateTime data = dataR.GetDateTime(10);
                        Trip trip = new Trip(place, transport, data, price, tickets, freeTickets);
                        trip.ID = idT;
                        Reservation reservation = new Reservation(client, telefon, trip, tickReserved);
                        reservation.ID = idR;
                        log.InfoFormat("Exiting findOne with value {0}", reservation);
                        return reservation;
                    }
                }
            }
            log.InfoFormat("Exiting findOne with value {0}", null);
            return null;
        }

        public IEnumerable<Reservation> findReservationByClient(string client1)
        {
            IDbConnection con = DBUtils.getConnection();
            log.InfoFormat("Finding reservation by client name: {0}",client1);
            IList<Reservation> resList = new List<Reservation>();
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select a.idr,a.client,a.telefon,a.tickets,b.idT,b.place,b.transport,b.price,b.tickets,b.freetickets,b.datat " +
                    "from reservation a " +
                    "inner join trip b " +
                    "on a.tripid=b.idT " +
                    "where a.client like @client";

                IDbDataParameter paramClient = comm.CreateParameter();
                paramClient.ParameterName = "@client";
                paramClient.Value = "%"+client1+"%";
                comm.Parameters.Add(paramClient);

                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        long idR = dataR.GetInt64(0);
                        string client = dataR.GetString(1);
                        string telefon = dataR.GetString(2);
                        int tickReserved = dataR.GetInt32(3);

                        long idT = dataR.GetInt64(4);
                        String place = dataR.GetString(5);
                        String transport = dataR.GetString(6);
                        double price = dataR.GetDouble(7);
                        int tickets = dataR.GetInt32(8);
                        int freeTickets = dataR.GetInt32(9);
                        DateTime data = dataR.GetDateTime(10);
                        Trip trip = new Trip(place, transport, data, price, tickets, freeTickets);
                        trip.ID = idT;
                        Reservation reservation = new Reservation(client, telefon, trip, tickReserved);
                        reservation.ID = idR;

                        resList.Add(reservation);
                    }
                }
            }

            return resList;
        }

        public IEnumerable<Reservation> findReservationByTripID(long tripID)
        {
            IDbConnection con = DBUtils.getConnection();
            log.InfoFormat("Finding reservation by trip ID: {0}", tripID);
            IList<Reservation> resList = new List<Reservation>();
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select a.idr,a.client,a.telefon,a.tickets,b.idT,b.place,b.transport,b.price,b.tickets,b.freetickets,b.datat " +
                    "from reservation a " +
                    "inner join trip b " +
                    "on a.tripid=b.idT " +
                    "where b.idT = @id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = tripID;
                comm.Parameters.Add(paramId);

                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        long idR = dataR.GetInt64(0);
                        string client = dataR.GetString(1);
                        string telefon = dataR.GetString(2);
                        int tickReserved = dataR.GetInt32(3);

                        long idT = dataR.GetInt64(4);
                        String place = dataR.GetString(5);
                        String transport = dataR.GetString(6);
                        double price = dataR.GetDouble(7);
                        int tickets = dataR.GetInt32(8);
                        int freeTickets = dataR.GetInt32(9);
                        DateTime data = dataR.GetDateTime(10);
                        Trip trip = new Trip(place, transport, data, price, tickets, freeTickets);
                        trip.ID = idT;
                        Reservation reservation = new Reservation(client, telefon, trip, tickReserved);
                        reservation.ID = idR;

                        resList.Add(reservation);
                    }
                }
            }

            return resList;
        }

        public void save(Reservation entity)
        {
            var con = DBUtils.getConnection();
            log.InfoFormat("Saving reservation = {0}", entity);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "insert into reservation (client,telefon,tripid,tickets) values (@client, @telefon,@tripId,@tick);select LAST_INSERT_ID()";
                /*var paramId = comm.CreateParameter();
                paramId.ParameterName = "@idf";
                paramId.Value = entity.ID;
                comm.Parameters.Add(paramId);*/

                var param1 = comm.CreateParameter();
                param1.ParameterName = "@client";
                param1.Value = entity.Client;
                comm.Parameters.Add(param1);

                var param2 = comm.CreateParameter();
                param2.ParameterName = "@telefon";
                param2.Value = entity.Telefon;
                comm.Parameters.Add(param2);

                var param3 = comm.CreateParameter();
                param3.ParameterName = "@tripid";
                param3.Value = entity.Trip.ID;
                comm.Parameters.Add(param3);

                var param4 = comm.CreateParameter();
                param4.ParameterName = "@tick";
                param4.Value = entity.NrTickets;
                comm.Parameters.Add(param4);

                

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        long idf = dataR.GetInt64(0);

                        Console.WriteLine("LAST ID INSERTED: {0}", idf);
                        entity.ID = idf;
                        log.InfoFormat("Reservation was added succesfully}");
                    }
                }

            }
        }

        public void update(Reservation entity)
        {
            var con = DBUtils.getConnection();
            log.InfoFormat("Updating agent = {0}", entity);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "update reservation set client=@client,telefon=@telefon,tripid=@tripid,tickets=@tick  where idr=@id";
                var paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = entity.ID;
                comm.Parameters.Add(paramId);

                var param1 = comm.CreateParameter();
                param1.ParameterName = "@client";
                param1.Value = entity.Client;
                comm.Parameters.Add(param1);

                var param2 = comm.CreateParameter();
                param2.ParameterName = "@telefon";
                param2.Value = entity.Telefon;
                comm.Parameters.Add(param2);

                var param3 = comm.CreateParameter();
                param3.ParameterName = "@tripid";
                param3.Value = entity.Trip.ID;
                comm.Parameters.Add(param3);

                var param4 = comm.CreateParameter();
                param4.ParameterName = "@tick";
                param4.Value = entity.NrTickets;
                comm.Parameters.Add(param4);

                var dataR = comm.ExecuteNonQuery();

                if (dataR == 0)
                    throw new RepositoryException("No reservation updated!");
                else
                    log.Info("Reservation was updated succesfully");

            }
        }
    }
}
